Implement floating point conversion with ryu

la-wu opened the pull request:
This PR implements floatDec and doubleDec using the Ryu algorithm described in Ulf Adams' paper. Compared to https://github.com/haskell/bytestring/pull/222, this is a native Haskell implementation based on https://github.com/Lumaere/ryu and the performance difference is relatively minimal.

However, this does add dependencies on template-haskell (which can probably be removed but will result in the hardcoding of large tables) and array.

Relevant benchmarks for this implementation on my i7-8700k
```
      foldMap floatDec (10000):                               OK (0.28s)
        1.0 ms ±  42 μs
      foldMap doubleDec (10000):                              OK (0.15s)
        1.1 ms ±  91 μs
```

And current
```
      foldMap floatDec (10000):                               OK (0.15s)
        9.5 ms ± 861 μs
      foldMap doubleDec (10000):                              OK (0.12s)
         17 ms ± 1.6 ms
```

Bodigrim commented:
Wonderful! 

We will drop older GHCs soon (see #265), do not worry about correspondent CI failures. 

Could you please submit tests as a separate PR?

Lumaere commented:
Moving off of those older GHCs is great. Any thoughts on the windows build failure
```
C:\Users\RUNNER~1\AppData\Local\Temp\ghcCE41.o.tmp: renameFile:renamePath:MoveFileEx "\\\\?\\C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\ghcCE41.o.tmp" Just "\\\\?\\C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\ghcCE41.o": does not exist (The system cannot find the file specified.)
```
Some light searching found https://gitlab.haskell.org/ghc/ghc/-/issues/18274 which suggests that there is some issue with `process` but I'm not very familiar with windows builds

sjakobi commented:
> Some light searching found https://gitlab.haskell.org/ghc/ghc/-/issues/18274 which suggests that there is some issue with `process` but I'm not very familiar with windows builds

IIUC we could work around this by using a different `cabal` executable. Maybe we could simply upgrade to using `cabal-install-3.4`?! Unfortunately I'm not very familiar with our use of GitHub Actions.

Bodigrim commented:
An issue with a temporary file used to happen before, but became much more prominent when Windows build switched to GHC 9.0. FIled as https://github.com/haskell/actions/issues/36. 

Not much can be done on our side, I believe, except just rerunning tests twice or thrice `cabal test || cabal test || cabal test`. I'll raise a PR.

Bodigrim commented:
Let's wait for https://github.com/haskell/actions/pull/39 to land, maybe it will alleviate our issues with Windows build.

Bodigrim commented:
> However, this does add dependencies on template-haskell (which can probably be removed but will result in the hardcoding of large tables) and array.

`template-haskell` is OK, #330 adds it to dependencies as well. 
`array` could be OKish, but I'd prefer to avoid. Could you please point me where is it used? I wonder if we can use [`Addr#`](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/exts/magic_hash.html#the-magic-hash) instead. Not only it eliminates a dependency, but also is usually significantly faster.

Lumaere commented:
The precomputed tables of multipliers and the formatting code both use arrays. Since they aren't exposed anywhere, I'll take a look at converting to raw address lookups

Lumaere commented:
@Bodigrim the `array` dependency has been removed but the build ran into the windows issue again.

Bodigrim commented:
@Lumaere I rerun, all green now :)

I'm a bit confused: f65dde31151c526725d1d4af27465f05d425d537 had tests, but they disappeared after push force. Could you please submit tests as a separate PR to be merged before this one?

Lumaere commented:
Can do. Some of the tests are more specific to internals (checking e.g round-trip and rounding). Should I stub those to the reference implementations or bring them back to this PR?

hsyl20 commented:
> template-haskell is OK, #330 adds it to dependencies as well.

Is it OK? `bytestring` is built with the stage1 compiler which doesn't support template-haskell.

Bodigrim commented:
@hsyl20 Thanks, good to know. How does `template-haskell` work for other boot libraries? I see that `text` and `exceptions` depend on it. 

We can review this with `template-haskell`, and replace it with generated splices as the very last step.

hsyl20 commented:
@Bodigrim The issue is that we can't link/execute generated code with GHC stage1, hence no TH splice.

* `exceptions` depends on `template-haskell` package but only to define `MonadThrow` for the `Q` monad, i.e. it doesn't use the TemplateHaskell extension
* `text` only uses TH quotes to define the `Lift` instances for `{Lazy.}Text`.

--------

> We can review this with template-haskell, and replace it with generated splices as the very last step.

Indeed. Or perhaps write a little utility program to generate the tables?

Also I think we could improve the generated tables by avoiding creating `ByteArray` at runtime. We can't create static `ByteArray` but we could generate primitive strings (which have type `Addr#`) and directly use `indexWord64OffAddr#`. For example `double_pow5_inv_split` would look like

```haskell
double_pow5_inv_split :: Addr#
double_pow5_inv_split = "\0\1\123\456..."#
```

We have to be careful about endianness though. Perhaps always generate little-endian tables and perform a byte-swap at runtime if the host is big-endian.

Bodigrim commented:
> Can do. Some of the tests are more specific to internals (checking e.g round-trip and rounding). Should I stub those to the reference implementations or bring them back to this PR?

@Lumaere sorry, missed your question. Do you mean that some tests from your branch fail under existing, `show`-based implementation (because of its infelicities)? In such case still submit them, wrapped into [`expectFail`](http://hackage.haskell.org/package/tasty-expected-failure-0.12.3/docs/Test-Tasty-ExpectedFailure.html#v:expectFail).

I'm keen to get this merged into the next release (which is September).

Lumaere commented:
@Bodigrim The tests were largely ported from the original C implementation of Ryu by Ulf Adams. I made the modifications to them so that they agree with `show` based implementation regarding rounding, etc. However, they still look at a new internal function (exposed for testing... they don't handle precision for all formats) called [`formatFloat` / `formatDouble`](https://github.com/haskell/bytestring/pull/365/files#diff-3016af887ac9c7d42c56d7358eacf8168698724938c79cf8239cdd4957eaccdaR27-R59) that is based loosely on [`formatRealFloat`](https://hackage.haskell.org/package/base-4.15.0.0/docs/GHC-Float.html#v:formatRealFloat). I can make the test PR expose these as
```
formatFloat :: FFFormat -> Maybe Int -> Float -> Builder
formatFloat fmt prec = string7 . formatRealFloat fmt prec

formatDouble :: FFFormat -> Maybe Int -> Double -> Builder
formatDouble fmt prec = string7 . formatRealFloat fmt prec
```
or modify all the tests to only use `floatDec` / `doubleDec`

Bodigrim commented:
From my perspective, it is better to modify tests to use only `floatDec` / `doubleDec`.

Lumaere commented:
Works for me. The PR is https://github.com/haskell/bytestring/pull/402. Let me know if you wanted it in a different format. Thanks

Bodigrim commented:
@la-wu could you please rebase?

Bodigrim commented on Data/ByteString/Builder/ASCII.hs:
```diff
@@ -35,9 +35,6 @@ module Data.ByteString.Builder.ASCII
     , wordDec
 
-    , floatDec
-    , doubleDec
```
This is a public module, we'd rather keep its API intact. Please import `Data.ByteString.Builder.RealFloat` and re-export `floatDec` / `doubleDec` from it.

Bodigrim commented on Data/ByteString/Builder/ASCII.hs:
```diff
@@ -163,22 +160,6 @@ wordDec :: Word -> Builder
--- | /Currently slow./ Decimal encoding of an IEEE 'Double'.
-{-# INLINE doubleDec #-}
-doubleDec :: Double -> Builder
-doubleDec = string7 . show
```
`string7` is defined above, but no longer used. Please remove it.

Bodigrim commented on Data/ByteString/Builder/RealFloat/Internal.hs:
```diff
@@ -0,0 +1,438 @@
+import Foreign.Ptr (castPtr)
+import Foreign.Storable (peek)
+{-# NOINLINE castFloatToWord32 #-}
+castFloatToWord32 :: Float -> Word32
```
Please add a comment to this function.

Bodigrim commented on Data/ByteString/Builder/RealFloat/Internal.hs:
```diff
@@ -0,0 +1,438 @@
+castFloatToWord32 :: Float -> Word32
+castFloatToWord32 x = unsafePerformIO (with x (peek . castPtr))
+
+-- | Convert a 'Double' to a 'Word64'.
```
```suggestion
-- | Interpret binary content of 'Double' as 'Word64', e. g., 'castDoubleToWord64' 1.0 = 4607182418800017408.
```
"Convert" is a bit ambiguous

Bodigrim commented on Data/ByteString/Builder/RealFloat/Internal.hs:
```diff
@@ -0,0 +1,438 @@
+asWord :: Integral w => Bool -> w
+asWord = fromIntegral . fromEnum
+
+pmap :: (a -> c) -> (a, b) -> (c, b)
```
This is just [`Control.Arrow.first`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Control-Arrow.html#v:first)

Bodigrim commented on Data/ByteString/Builder/RealFloat/Internal.hs:
```diff
@@ -0,0 +1,438 @@
+
+{-# INLINABLE (.<<) #-}
+(.<<) :: (Bits a, Integral b) => a -> b -> a
+a .<< s = unsafeShiftL a (fromIntegral s)
```
To be honest, I'd advise against such, double-polymorphic functions. They tend to make code difficult to validate and error-prone when refactoring. Is it possible to limit them to several specific instantiations?

la-wu replied:
Do you think it would be sufficient to make them single-polymorphic e.g `(Bits a, Integral a) => a -> a -> a`?

Bodigrim replied:
Yes, making them single-polymorphic sounds good.

Bodigrim commented on Data/ByteString/Builder/RealFloat/Internal.hs:
```diff
@@ -0,0 +1,438 @@
+--    doubles: 1 (sign) + 17 (mantissa) + 1 (.) + 1 (e) + 4 (exponent) = 24
+--
+maxEncodedLength :: Int
+maxEncodedLength = 32
```
So why it it 32 instead of 24?

la-wu replied:
IIRC, I liked rounding to the nearest power of 2 for whatever reason

Bodigrim commented on Data/ByteString/Builder/RealFloat/Internal.hs:
```diff
@@ -0,0 +1,438 @@
+boundString s = boundedPrim maxEncodedLength $ const (pokeAll s)
+
+--         Sign -> Exp  -> Mantissa
+special :: Bool -> Bool -> Bool -> BoundedPrim ()
```
Could we possibly define `data Foo = Foo { fooSign :: !Bool, fooExpr :: !Bool, fooMantissa :: !Bool }`, so that the meaning of arguments is self-explanatory, and their ordering is less error-prone?

la-wu replied:
Is https://github.com/haskell/bytestring/pull/365/files#diff-e91607876ac26bb3b25c8b060668d06eb50becfd5a682bd59636e8de5bb9f0d3R214 close enough?

Bodigrim commented on Data/ByteString/Builder/RealFloat/Internal.hs:
```diff
@@ -0,0 +1,438 @@
+toAscii :: Word# -> Word#
+toAscii a = a `plusWord#` asciiRaw asciiZero
+
+data ByteArray = ByteArray (ByteArray#)
```
```suggestion
data ByteArray = ByteArray ByteArray#
```

Bodigrim commented on Data/ByteString/Builder/RealFloat/Internal.hs:
```diff
@@ -0,0 +1,438 @@
+
+-- Returns floor(log_10(2^e)); requires 0 <= e <= 1650.
+log10pow2Unboxed :: Int# -> Int#
+log10pow2Unboxed e = (e *# 78913#) `uncheckedIShiftRL#` 18#
```
These magic numbers could benefit from an explanation in haddock, explaining that `78913 / 2^18` is a sufficiently good approximation of `logBase 10 2`.

Bodigrim commented on Data/ByteString/Builder/RealFloat/Internal.hs:
```diff
@@ -0,0 +1,438 @@
+-- acceptBoundsUnboxed v = ((v `uncheckedShiftRL#` 2#) `and#` 1##) `eqWord#` 0##
+
+fquot10 :: Word# -> Word#
+fquot10 w = (w `timesWord#` 0xCCCCCCCD##) `uncheckedShiftRL#` 35#
```
Magic constants for division here and below require either an explanation why it works, or a reference to some source. For instance, I'd point readers to `https://hackage.haskell.org/package/quote-quot-0.2.0.0/docs/Numeric-QuoteQuot.html#v:astQuot` as a mean of verification.

Bodigrim reviewed (commented):
@la-wu thanks for your astonishing work! It will take some time to review, here are my first suggestions.

sjakobi reviewed (commented):
First off, apologies for the late review! I hadn't realized that this PR had been waiting for a review for so long.

I've made a first pass over the code. So far I'm pretty impressed by the clean and well-documented internals! :+1:

I should point out that I'm not very experienced with floating-point encodings, so I probably won't be able to comment on much of the internals. For correctness, I'll have to rely on the test suite, which I'm grateful you have improved so much! :)

I have a few requests:

* [ ] Please add module header haddocks to all the added modules.
* [ ] Since I have encountered several bugs that involved `fromIntegral`, I have a somewhat irrational fear of it. (I also dislike that I often can't tell the types involved just from looking at it.) Would it be possible to replace its use with specialized `intToFloat`-style helpers? Alternatively, we could consider using `-XTypeApplications`, but I think that would require dropping support for GHC < 8 a bit earlier than planned so far.
* [ ] I would also prefer if the types `(.>>)` and `(.<<)` were constrained a bit, if that can be done ergonomically.

la-wu commented:
Thanks for the review. If you're interested in reading about the floating point string conversion, I think part 2 of [the ryu paper](https://dl.acm.org/doi/pdf/10.1145/3192366.3192369) has a good intro to it.

>     * [ ]  Please add module header haddocks to all the added modules.

Will do. What do you suggest I put in the fields Copyright, Maintainer, Stability?

On stability (and wrt some of your other comments that I'll get to individually as I start making edits), perhaps separating the current `RealFloat.hs` into a stable portion for just `floatDec / doubleDec` and a internal/experimental portion that has the `formatFloat / formatDouble` (which are also new to bytestring) could be clearer. What are your thoughts?

>     * [ ]  Since I have encountered several bugs that involved `fromIntegral`, I have a somewhat irrational fear of it. (I also dislike that I often can't tell the types involved just from looking at it.) Would it be possible to replace its use with specialized `intToFloat`-style helpers? Alternatively, we could consider using `-XTypeApplications`, but I think that would require dropping support for GHC < 8 a bit earlier than planned so far.
> 
>     * [ ]  I would also prefer if the types `(.>>)` and `(.<<)` were constrained a bit, if that can be done ergonomically.

I think these can both be at least partially achieved. Most (all?) of the `fromIntegral` usages are actually just signed-to-unsigned or 32-to-64-bit conversion (and vice versa). Some of it can be elided by a more careful handling of integral types and others can get explicit annotations of the whole expression (like in `Builder/Prim/Binary.hs`)

sjakobi commented:
> Will do. What do you suggest I put in the fields Copyright, Maintainer, Stability?

Oh, I was mostly looking for brief module descriptions. I'm not sure we need all that other data. I usually skip it in other packages that I maintain. (CC @Bodigrim)

Do feel free to add your copyright in the modules that you've written of course!

> On stability (and wrt some of your other comments that I'll get to individually as I start making edits), perhaps separating the current `RealFloat.hs` into a stable portion for just `floatDec / doubleDec` and a internal/experimental portion that has the `formatFloat / formatDouble` (which are also new to bytestring) could be clearer. What are your thoughts?

I do think that `formatFloat / formatDouble` are very desirable additions. If there is particular reason to believe that these functions will change in the near future, it might be better to to export them from an `Internal` module, so we don't have to do a major version bump. `D.B.B.RealFloat.Internal` would be the obvious candidate, but then you'd probably want to move the other contents of that module elsewhere…

I'm not overly concerned about breaking changes though. As long as there's a general sense of "improvement", I think we can do a major version bump every 2 or 3 years.

sjakobi commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,137 @@
+
+module Data.ByteString.Builder.RealFloat
+  ( FFFormat(..)
```
As suggested in https://gitlab.haskell.org/ghc/ghc/-/issues/20382#note_376486, I think we should define our own format type, if that can be done without major downsides.

la-wu replied:
That's fine but it just introduces another name to know for using the API (divergent from `GHC.Float.formatRealFloat`). Do you have a preference on the name?

sjakobi replied:
> Do you have a preference on the name?

Just `Format`?! `FloatFormat`?!

Bodigrim commented:
> What do you suggest I put in the fields Copyright, Maintainer, Stability?

Feel free to put your own Copyright to new modules. Leave Maintainer empty or put yourself if you wish. Ignore Stability field.

> Since I have encountered several bugs that involved fromIntegral, I have a somewhat irrational fear of it. (I also dislike that I often can't tell the types involved just from looking at it.) Would it be possible to replace its use with specialized intToFloat-style helpers? 

I very much second this. `fromIntegral` is extremely dangerous, let's replace it with monomorphic helpers, similar to https://github.com/haskell/text/commit/3bcde5e3aa1bb78fad3ed7e5335f8891a4a3bdc8

@la-wu would you like to take a stab at eliminating TemplateHaskell in this PR? Ideally we would like to end up with pregenerated lookup tables similar to  https://github.com/haskell/bytestring/blob/688f3c0887f2ca0623f2f54f78e8f675f92e31bf/Data/ByteString/Builder/Prim/Internal/Base16.hs

la-wu commented:
> I very much second this. `fromIntegral` is extremely dangerous, let's replace it with monomorphic helpers, similar to [haskell/text@3bcde5e](https://github.com/haskell/text/commit/3bcde5e3aa1bb78fad3ed7e5335f8891a4a3bdc8)

In https://github.com/haskell/bytestring/pull/365/commits/c54cffc85041f78fce60ce9694d5a184b611d7cf I was able to remove most of the `fromIntegral` usages and annotated the remainder with expected types of the resulting conversion. Does this seem sufficient or should the source type also be explicit?

> @la-wu would you like to take a stab at eliminating TemplateHaskell in this PR? Ideally we would like to end up with pregenerated lookup tables similar to https://github.com/haskell/bytestring/blob/688f3c0887f2ca0623f2f54f78e8f675f92e31bf/Data/ByteString/Builder/Prim/Internal/Base16.hs

I certainly can just pre-generate the tables and paste them in the source but they are relatively large (double is ~650 entries total) https://github.com/la-wu/ryu/commit/13ed23f003174a1d4547e456cf08a810357e5c1f#diff-c1a8cf82fcd60539caf14dd3eb1d6c1c4d3a909ad823e525f03636b4e8acd73bL95

Bodigrim commented:
> Does this seem sufficient or should the source type also be explicit?

Let's eliminate all calls to `fromIntegral` and make source types explicit as well. 

> I certainly can just pre-generate the tables and paste them in the source but they are relatively large (double is ~650 entries total)

As long as you retain the code used for generation in comments nearby, it's fine. Please use `Addr#` instead of `UArray`.

Bodigrim commented:
@la-wu unfortunately this fails on big endian arch. You can test it as follows:

```
apt install -y docker.io && \
docker run --rm --privileged multiarch/qemu-user-static --reset -p yes && \
docker run --rm -i -t s390x/ubuntu
```
and then inside the virtual machine
```
apt update && \
apt install -y ghc cabal-install git build-essential && \
git clone https://github.com/la-wu/bytestring && \
cd bytestring && \
git checkout ryu-native && \
cabal new-update && \
cabal -j1 new-test test-builder
```

```
All
  Data.ByteString.Builder
    toLazyByteStringWith:       FAIL (0.01s)
      *** Failed! Falsified (after 2 tests):
      recipe: Recipe Untrimmed 0 33336 "" [FDec 0.95366085]
      render: 0.95366085
      build : 1.0e-8
      diff  : [(48,49),(46,46),(57,48),(53,101),(51,45),(54,56)]

      Use -p '/toLazyByteStringWith/' to rerun this test only.
    runBuilder:                 FAIL
      *** Failed! Falsified (after 3 tests and 1 shrink):
      [FDec (-1.0)]
      Use -p '/runBuilder/' to rerun this test only.
    floatDec:                   FAIL
      *** Failed! Falsified (after 2 tests and 5 shrinks):
      -0.1
      Use -p '/floatDec/' to rerun this test only.
    doubleDec:                  FAIL (0.01s)
      *** Failed! Falsified (after 2 tests and 4 shrinks):
      0.1
      Use -p '/doubleDec/' to rerun this test only.
    f2sBasic:                   FAIL
      *** Failed! Falsified (after 1 test):
      Use -p '/f2sBasic/' to rerun this test only.
    f2sSubnormal:               FAIL (0.01s)
      *** Failed! Falsified (after 1 test):
      Use -p '/f2sSubnormal/' to rerun this test only.
    f2sMinAndMax:               FAIL (0.01s)
      *** Failed! Falsified (after 1 test):
      Use -p '/f2sMinAndMax/' to rerun this test only.
    f2sBoundaryRound:           FAIL (0.01s)
      *** Failed! Falsified (after 1 test):
      Use -p '/f2sBoundaryRound/' to rerun this test only.
    f2sExactValueRound:         FAIL
      *** Failed! Falsified (after 1 test):
      Use -p '/f2sExactValueRound/' to rerun this test only.
    f2sTrailingZeros:           FAIL (0.01s)
      *** Failed! Falsified (after 1 test):
      Use -p '/f2sTrailingZeros/' to rerun this test only.
    f2sRegression:              FAIL
      *** Failed! Falsified (after 1 test):
      Use -p '/f2sRegression/' to rerun this test only.
    f2sLooksLikePowerOf5:       FAIL (0.01s)
      *** Failed! Falsified (after 1 test):
      Use -p '/f2sLooksLikePowerOf5/' to rerun this test only.
    f2sOutputLength:            FAIL
      *** Failed! Falsified (after 1 test):
      Use -p '/f2sOutputLength/' to rerun this test only.
    d2sSubnormal:               FAIL
      *** Failed! Falsified (after 1 test):
      Use -p '/d2sSubnormal/' to rerun this test only.
    d2sMinAndMax:               FAIL
      *** Failed! Falsified (after 1 test):
      Use -p '/d2sMinAndMax/' to rerun this test only.
    d2sTrailingZeros:           FAIL
      *** Failed! Falsified (after 1 test):
      Use -p '/d2sTrailingZeros/' to rerun this test only.
    d2sRegression:              FAIL
      *** Failed! Falsified (after 1 test):
      Use -p '/d2sRegression/' to rerun this test only.
    d2sLooksLikePowerOf5:       FAIL
      *** Failed! Falsified (after 1 test):
      Use -p '/d2sLooksLikePowerOf5/' to rerun this test only.
    d2sOutputLength:            FAIL
      *** Failed! Falsified (after 1 test):
      Use -p '/d2sOutputLength/' to rerun this test only.
    d2sMinMaxShift:             FAIL
      *** Failed! Falsified (after 1 test):
      Use -p '/d2sMinMaxShift/' to rerun this test only.
    d2sSmallIntegers:           FAIL
      *** Failed! Falsified (after 1 test):
      Use -p '/d2sSmallIntegers/' to rerun this test only.
    f2sPowersOf10:              FAIL
      *** Failed! Falsified (after 1 test):
      Use -p '/f2sPowersOf10/' to rerun this test only.
    d2sPowersOf10:              FAIL (0.01s)
      *** Failed! Falsified (after 1 test):
      Use -p '/d2sPowersOf10/' to rerun this test only.
      
23 out of 256 tests failed (55.38s)
```

See suggestions in https://github.com/haskell/bytestring/pull/365#issuecomment-788795773

la-wu commented:
> See suggestions in [#365 (comment)](https://github.com/haskell/bytestring/pull/365#issuecomment-788795773)

I see. I forgot someone already thought about this ha. It seems like updating a primitive string causes a runtime exception. Are we OK with a runtime swap for each call on big-endian systems? From the `GHC.Prim` docs it seems like the other `*Array#` options are stored on the heap. Is that a preferable option to a primitive string?

Bodigrim commented:
Runtime swap in case of big endian is absolutely fine, please proceed with it. You can find a reliable method to detect endianness in the test suite.

la-wu commented:
Sounds good. I'm working on a laptop ATM but I'll do the big-endian testing when I get home later tonight.

Bodigrim commented on Data/ByteString/Builder/RealFloat/Internal.hs:
```diff
@@ -0,0 +1,732 @@
+{-# INLINE getWord64At #-}
+getWord64At :: Addr# -> Int# -> Word#
+getWord64At arr i =
+#if defined(WORDS_BIGENDIAN)
```
You need `#include <ghcautoconf.h>` somewhere for this to work.

Bodigrim commented:
@la-wu I retested on s390x: some tests are still failing:

```
All
  Data.ByteString.Builder
    toLazyByteStringWith:       FAIL (0.06s)
      *** Failed! Falsified (after 5 tests):
      recipe: Recipe Untrimmed 33332 33333 "\NUL\STX\EOT\EOT\STX\ETX\EOT" [ModState 4,String "",SBS Hex "\NUL",DDec 1.7874421410691919]
      render: 33332001.7874421410691919
      build : 33332001.0e-81
      diff  : [(55,48),(56,101),(55,45),(52,56),(52,49),(50,0),(49,2),(52,4),(49,4),(48,2),(54,3),(57,4)]

      Use --quickcheck-replay=170216 to reproduce.
      Use -p '/toLazyByteStringWith/' to rerun this test only.
    runBuilder:                 FAIL (0.02s)
      *** Failed! Falsified (after 7 tests and 4 shrinks):
      [DDec (-0.1)]
      Use --quickcheck-replay=146262 to reproduce.
      Use -p '/runBuilder/' to rerun this test only.
    doubleDec:                  FAIL
      *** Failed! Falsified (after 2 tests and 4 shrinks):
      -0.1
      Use --quickcheck-replay=630660 to reproduce.
      Use -p '/doubleDec/' to rerun this test only.
    f2sSubnormal:               FAIL
      *** Failed! Falsified (after 1 test):
      Use --quickcheck-replay=837181 to reproduce.
      Use -p '/f2sSubnormal/' to rerun this test only.
    f2sMinAndMax:               FAIL
      *** Failed! Falsified (after 1 test):
      Use --quickcheck-replay=323824 to reproduce.
      Use -p '/f2sMinAndMax/' to rerun this test only.
    f2sBoundaryRound:           FAIL
      *** Failed! Falsified (after 1 test):
      Use --quickcheck-replay=120739 to reproduce.
      Use -p '/f2sBoundaryRound/' to rerun this test only.
    f2sExactValueRound:         OK (0.01s)
      +++ OK, passed 1 test.
    f2sTrailingZeros:           FAIL
      *** Failed! Falsified (after 1 test):
      Use --quickcheck-replay=300349 to reproduce.
      Use -p '/f2sTrailingZeros/' to rerun this test only.
    f2sRegression:              FAIL
      *** Failed! Falsified (after 1 test):
      Use --quickcheck-replay=600654 to reproduce.
      Use -p '/f2sRegression/' to rerun this test only.
    f2sLooksLikePowerOf5:       FAIL
      *** Failed! Falsified (after 1 test):
      Use --quickcheck-replay=43709 to reproduce.
      Use -p '/f2sLooksLikePowerOf5/' to rerun this test only.
    f2sOutputLength:            FAIL
      *** Failed! Falsified (after 1 test):
      Use --quickcheck-replay=240087 to reproduce.
      Use -p '/f2sOutputLength/' to rerun this test only.
    d2sSubnormal:               FAIL
      *** Failed! Falsified (after 1 test):
      Use --quickcheck-replay=251723 to reproduce.
      Use -p '/d2sSubnormal/' to rerun this test only.
    d2sMinAndMax:               FAIL
      *** Failed! Falsified (after 1 test):
      Use --quickcheck-replay=218701 to reproduce.
      Use -p '/d2sMinAndMax/' to rerun this test only.
    d2sTrailingZeros:           FAIL (0.01s)
      *** Failed! Falsified (after 1 test):
      Use --quickcheck-replay=994330 to reproduce.
      Use -p '/d2sTrailingZeros/' to rerun this test only.
    d2sRegression:              FAIL
      *** Failed! Falsified (after 1 test):
      Use --quickcheck-replay=768544 to reproduce.
      Use -p '/d2sRegression/' to rerun this test only.
    d2sLooksLikePowerOf5:       FAIL
      *** Failed! Falsified (after 1 test):
      Use --quickcheck-replay=623221 to reproduce.
      Use -p '/d2sLooksLikePowerOf5/' to rerun this test only.
    d2sOutputLength:            FAIL
      *** Failed! Falsified (after 1 test):
      Use --quickcheck-replay=798900 to reproduce.
      Use -p '/d2sOutputLength/' to rerun this test only.
    d2sMinMaxShift:             FAIL (0.01s)
      *** Failed! Falsified (after 1 test):
      Use --quickcheck-replay=776656 to reproduce.
      Use -p '/d2sMinMaxShift/' to rerun this test only.
    d2sSmallIntegers:           FAIL
      *** Failed! Falsified (after 1 test):
      Use --quickcheck-replay=180148 to reproduce.
      Use -p '/d2sSmallIntegers/' to rerun this test only.
    f2sPowersOf10:              FAIL
      *** Failed! Falsified (after 1 test):
      Use --quickcheck-replay=488089 to reproduce.
      Use -p '/f2sPowersOf10/' to rerun this test only.
    d2sPowersOf10:              FAIL
      *** Failed! Falsified (after 1 test):
      Use --quickcheck-replay=455707 to reproduce.
      Use -p '/d2sPowersOf10/' to rerun this test only.

20 out of 256 tests failed (57.36s)
```

la-wu commented:
Sorry for the delay. I reproduced the issues on a docker image and was able to get those tests passing (more big-endian byte-order differences as expected).

Bodigrim commented:
@la-wu great! Could you please run `cabal test --enable-coverage test-builder`? Is it possible to add test cases, covering more lines?

la-wu commented:
I added a few more for tests and removed some now-unused table generation code.

There was also one bug caught from this where a rounding branch was never taken. This doesn't trip any test cases as far as I can tell since round-even is ignored in favor of matching `show`... but good catch nonetheless. The remaining unused alternatives / expressions all revolve around round-even not being used (specifically `acceptBoundsUnboxed` always returns False-y)

la-wu commented:
On my machine there seems to have been a small perf hit. Haven't dug into what caused it though
```
      foldMap floatDec (10000):           OK (0.16s)
        1.2 ms ± 104 μs
      foldMap doubleDec (10000):          OK (0.34s)
        1.3 ms ±  96 μs
```

hsyl20 commented on Data/ByteString/Builder/RealFloat/Internal.hs:
```diff
@@ -0,0 +1,758 @@
+    , vvIsTrailingZeros :: !Bool
+    }
+
+-- | Trim digits while and update bookkeeping state when the table-computed
```
"Trim digits while" what?

hsyl20 commented on Data/ByteString/Builder/RealFloat/Internal.hs:
```diff
@@ -0,0 +1,758 @@
+        !(vw', _    ) = quotRem10Boxed $ vw d
+
+
+-- | Trim digits while and update bookkeeping state when the table-computed
```
Same

la-wu commented:
@Bodigrim the breakage in 9.2.1 seems to be from a change in ghc-prim where the [wording in 9.0.1](https://downloads.haskell.org/~ghc/9.0.1/docs/html/libraries/ghc-prim-0.7.0/GHC-Prim.html) says

> The 32-bit sizes are represented using Int# and Word# when WORD_SIZE_IN_BITS $geq$ 32; otherwise, these are represented using distinct primitive types Int32# and Word32#

Versus the [wording in 9.2.1](https://downloads.haskell.org/~ghc/9.2.1-rc1/docs/html/libraries/ghc-prim-0.8.0/GHC-Prim.html#g:8) says

> The 64-bit sizes are represented using Int# and Word# when WORD_SIZE_IN_BITS $geq$ 64; otherwise, these are represented using distinct primitive types Int64# and Word64#

So it seems like implicitly I've been relying on `Word#` to really be 64-bit always (and something like `data Word32 = W32 Word#` whereas now it is `data Word32 = W32 Word32#`). I'll take a look at removing those assumptions when I get back.

Bodigrim commented:
@la-wu yeah, GHC 9.2 completely overhauls sub-word primitive types. I've added a CI configuration for GHC 9.2, please rebase on the latest `master` when back. 

Other than that the branch looks good to me, thanks!

la-wu commented:
@Bodigrim sorry for the delay. There were more changes than expected because this PR previously didn't work for 32-bit systems. I've checked locally now that tests work for 64-bit big- and little-endian systems and 32-bit LE.

sjakobi commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,179 @@
+  | FGeneric        -- ^ dispatches to fixed or exponent based on the exponent
+  deriving Show
+
+-- TODO: support precision argument for FGeneric and FExponent
```
@la-wu What's the reason why you left this for future work? Is it tricky to implement?

It's a bit of a wart that the `Maybe Int` argument is meaningless in the `FExponent` case. I was wondering whether we ought to hide the wart and improve the forward-compatibility by creating an opaque format type somewhat like this:

```haskell
data FloatFormat = MkFloatFormat
  { floatFormat' :: FloatFormat' -- the old FF type
  , precision :: Maybe Int
  }

fixed :: Int -> FloatFormat
fixed n = MkFloatFormat FFixed (Just n)

scientific :: FloatFormat
scientific = MkFloatFormat FExponent Nothing

-- added once the precision support is complete
scientificWithPrecision :: Int -> FloatFormat
scientific n = MkFloatFormat FExponent (Just n)

-- generic FloatFormats
...
```

la-wu replied:
It was mainly left unimplemented because we previously only had `floatDec` and `doubleDec` so I did not spend too much time on it. I implemented the `FFixed` case to match `show` and what we had before (we need to use non-exponent formatting for `e >= 0 && e <= 7`).

I don't think it would be particularly hard to implement something for precision for FExponent (we should be able to largely follow what is in base https://github.com/ghc/packages-base/blob/master/GHC/Float.lhs#L602 as I did for the fixed-format). But ideally we would move to using a raw buffer-formatter for both FFixed and FExponent with precision so they could be equally efficient there.

sjakobi replied:
@Bodigrim How do you feel about future-proofing the API along the lines I suggested above?

An additional advantage of an opaque type is that I think we might get away with not providing any instances for it. With public constructors, I think we ought to provide `Show`, `Ord`, `Enum` etc.

Bodigrim replied:
I think an opaque type is a good idea.

la-wu replied:
Took a stab at this. Is this what you guys had in mind? https://github.com/haskell/bytestring/pull/365/commits/0474332d4fca7f757e024af5f2b6449b73d04672

la-wu commented:
Sorry, I've been busy but I'll try to make some time this weekend to review the latest comments.

la-wu commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,179 @@
+doubleDec = formatDouble FGeneric Nothing
+
+-- | ByteString float-to-string format
+data FloatFormat
+  = FExponent       -- ^ scientific notation
+  | FFixed          -- ^ fixed precision with `Maybe Int` digits after the decimal
+  | FGeneric        -- ^ dispatches to fixed or exponent based on the exponent
```
I don't mind removing it. We were advised previously to not depend on the `GHC.Float` type [FFFormat](https://hackage.haskell.org/package/base-4.16.0.0/docs/GHC-Float.html#t:FFFormat) so I this one.

la-wu commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,179 @@
+
+-- | ByteString float-to-string format
+data FloatFormat
+  = FExponent       -- ^ scientific notation
```
This was modeled after the `GHC.Float` type [FFFormat](https://hackage.haskell.org/package/base-4.16.0.0/docs/GHC-Float.html#t:FFFormat) but I'm happy to change it

sjakobi replied:
I think this `FFFormat` type isn't very well known, so our users are unlikely to be familiar with it.

Please change the name to `FScientific`

sjakobi commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,200 @@
+data FloatFormat
+  = FExponent       -- ^ scientific notation
+  | FFixed          -- ^ fixed precision with `Maybe Int` digits after the decimal
+  | FGeneric        -- ^ dispatches to fixed or exponent based on the exponent
+  deriving Show
```
```suggestion
  | FGeneric        -- ^ dispatches to fixed precision or scientific notation based on the exponent
  deriving Show
```

sjakobi commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,227 @@
+-- | An opaque wrapper around `FloatFormat'` and an optional precision argument
+-- used by `formatFloat` and `formatDouble`.
+data FloatFormat = MkFloatFormat
+  { floatFormat' :: FloatFormat'
+  , precision :: Maybe Int
```
If these record selectors aren't needed, better remove them to avoid the warnings.

sjakobi replied:
Also, should the fields be strict (`!`)?! I'm not sure.

la-wu replied:
I left them as non-strict but removed the records. I didn't think too hard about it but in theory if the format gets inlined it shouldn't matter?

sjakobi commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,227 @@
+doubleDec :: Double -> Builder
+doubleDec = formatDouble generic
+
+-- | An opaque wrapper around `FloatFormat'` and an optional precision argument
+-- used by `formatFloat` and `formatDouble`.
```
This documentation is primarily for the API user who doesn't need to be aware of the internals. Just say something like

```suggestion
-- | Format type for use with `formatFloat` and `formatDouble`.
```

sjakobi commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,227 @@
+-- Mentions of 'shortest' in the docs below are with this caveat.
+
+module Data.ByteString.Builder.RealFloat
+  ( FloatFormat
+  , floatDec
+  , doubleDec
+  , fixed
+  , fixedDefaultPrecision
+  , scientific
+  , generic
+  , formatFloat
+  , formatDouble
```
```suggestion
  ( floatDec
  , doubleDec
  
  -- * Custom formatting
  , formatFloat
  , formatDouble
  , FloatFormat
  , fixed
  , fixedDefaultPrecision
  , scientific
  , generic
```

sjakobi commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,227 @@
+import GHC.Word (Word64)
+import GHC.Show (intToDigit)
+
+-- | Returns a rendered Float. Matches `show` in displaying in fixed or
+-- scientific notation
```
```suggestion
-- | Returns a rendered Float. Matches `show` in displaying in fixed or
-- scientific notation
--
-- @
-- floatDec = 'formatFloat' 'generic'
-- @
```

sjakobi commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,227 @@
+fixed :: Int -> FloatFormat
+fixed n = MkFloatFormat FFixed (Just n)
+
+-- | Standard notation with the default precision (number of decimal places)
+fixedDefaultPrecision :: FloatFormat
+fixedDefaultPrecision = MkFloatFormat FFixed Nothing
```
The number of the decimal places isn't truly "fixed", right? Maybe call it `decimal`?!

la-wu replied:
`fixed` is from `FFFormat`. I renamed it all to `standard` notation but am open to alternatives if there's a more canonical name

sjakobi replied:
I was hoping to find a canonical name for this format but couldn't find one. `standard` seems good enough. If we find a better name later on, it would be easy to add it and to deprecate the old one.

sjakobi commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,227 @@
+generic = MkFloatFormat FGeneric Nothing
+
+-- | ByteString float-to-string format
+data FloatFormat'
```
A name more distinct from `FloatFormat` would probably aid readability

```suggestion
data FormatMode
```

sjakobi reviewed (commented):
Thanks, @la-wu. I like this direction! Needs just a bit more polishing.

sjakobi commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,227 @@
+-- where the closest and 'shortest' string representations are not used.
+-- Mentions of 'shortest' in the docs below are with this caveat.
+
+module Data.ByteString.Builder.RealFloat
```
I wonder whether there's a better module name for this module. Is there a good motivation for naming it `D.B.B.RealFloat`? If it's about the instances of `RealFloat`, why not name it `D.B.B.RealFrac` instead?! `RealFrac` has basically the same instances after all.

Or more seriosly: `D.B.B.Floats`?!

la-wu replied:
I think this is another artifact of `GHC.Float` and `format**RealFloat**`. Looking at the classes at least though, we do depend on `floatRange` and `floatDigits` from `RealFloat` (though these could be relatively trivially made constants as we have for several other float and double bitcounts)

sjakobi replied:
I'm fine with sticking with this name for now – I don't have a much better suggestion at least.

Thoughts, @Bodigrim?

Bodigrim replied:
I'm fine with this name.

Bodigrim commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,227 @@
+-- "12.345"
+{-# INLINABLE formatFloat #-}
+formatFloat :: FloatFormat -> Float -> Builder
+formatFloat (MkFloatFormat fmt prec) f =
```
Yes, it will be better for inlining this way.

la-wu replied:
I updated this. Might there be problems with assuming https://github.com/haskell/bytestring/pull/365/files#diff-3016af887ac9c7d42c56d7358eacf8168698724938c79cf8239cdd4957eaccdaR114 is lazy?

sjakobi replied:
Can you clarify what kind of problems you're concerned about? The easiest thing might be to take a look at the Core…

la-wu replied:
Just that running `f2Intermediate` is unnecessary work in case of `FScientific`. I took a look at the `dump-simpl` output but it wasn't obvious to me what I should be looking for...

sjakobi commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,231 @@
+-- Floating point formatting for Bytestring.Builder
+--
+-- This module primarily exposes `floatDec` and `doubleDec` which do the
+-- equivalent of converting through `string7 . show`.
```
```suggestion
-- equivalent of converting through @'string7' . 'show'@.
```

sjakobi commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,231 @@
+-- It also exposes `formatFloat` and `formatDouble` with a similar API as
+-- `GHC.Float.formatRealFloat`.
+--
+-- NB: this implementation matches `show`'s output (specifically with respect
+-- to default rounding and length). In particular, there are boundary cases
+-- where the closest and 'shortest' string representations are not used.
+-- Mentions of 'shortest' in the docs below are with this caveat.
```
This paragraph is a bit hard to understand I think. An example would be very helpful. Please also clarify what "this implementation" refers to. Maybe create a [named chunk](https://haskell-haddock.readthedocs.io/en/latest/markup.html#named-chunks-of-documentation) and refer to it from the relevant places.

Also note that single quotes (`'`) are used for marking identifiers in Haddock. Escaping the quotes might work?! E.g. `\'shortest\'`.

la-wu replied:
I added an example / handwavy explanation about the rounding / boundaries and pointed to the paper for details. TBH at this point I am far enough from the writing this code that I would need to redigest the exact semantics.

WRT quotes, thanks for pointing that out. I'm used to normal MD.

sjakobi commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,231 @@
+standard :: Int -> FloatFormat
+standard n = MkFloatFormat FStandard (Just n)
+
+-- | Standard notation with the default precision (number of decimal places)
```
"default precision" seems like another term that could be explained in a named chunk that can be referenced from here.

la-wu replied:
It just means matching `show` / `formatRealFloat`. LMK if I should elaborate further...

sjakobi reviewed (commented):
Just a few more documentation tweaks.

Could you possibly generate the haddocks and check whether the markup comes out right? Especially that identifiers are hyperlinked and non-identifiers aren't accidentally rendered in mono-space font?

Bodigrim commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,269 @@
+           in mk0 ls `mappend` mkDot rs
+      | otherwise ->
+          let (ei, is') = roundTo 10 p' (replicate (-e) 0 ++ ds)
+              (b:bs) = digitsToBuilder (if ei > 0 then is' else 0:is')
```
This raises `Pattern match(es) are non-exchaustive` with GHC 9.2.

la-wu replied:
Hm is there a way to convince GHC that the list is non-empty? Looking at [the implementation of `roundTo`](https://hackage.haskell.org/package/base-4.16.0.0/docs/src/GHC.Float.html#roundTo), the non-throwing return values are `(0, _)` and `(1, 1:_)`. Before the call to `digitsToZero`, we prepend `0` to the former case's `snd` value so AFIACT this is 'correct'.

la-wu replied:
We could also just mimic the case above or add some redundant matches e.g
```
diff --git a/Data/ByteString/Builder/RealFloat.hs b/Data/ByteString/Builder/RealFloat.hs
index cd0d39d..0dadf45 100644
--- a/Data/ByteString/Builder/RealFloat.hs
+++ b/Data/ByteString/Builder/RealFloat.hs
@@ -258,8 +258,8 @@ showStandard m e prec =
            in mk0 ls `mappend` mkDot rs
       | otherwise ->
           let (ei, is') = roundTo 10 p' (replicate (-e) 0 ++ ds)
-              (b:bs) = digitsToBuilder (if ei > 0 then is' else 0:is')
-           in b `mappend` mkDot bs
+              (ls, rs) = splitAt 1 $ digitsToBuilder (if ei > 0 then is' else 0:is')
+           in mk0 ls `mappend` mkDot rs
           where p' = max p 0
   where
     mk0 ls = case ls of [] -> char7 '0'; _ -> mconcat ls
```

Bodigrim replied:
I'm not saying that your code is wrong, I'm saying that we need to get rid of a warning. Boot packages cannot have warnings. Throwing an `error` for an empty list would be fine, for instance.

sjakobi replied:
I agree that this warning needs to be silenced. Maybe this would be a good fit for `Data.List.NonEmpty`?!

la-wu replied:
I added the snippet above. A lot of the code in `GHC.Float` makes the same assumption though so do they just turn the warning off?

sjakobi replied:
Yes. That module uses the pragma

```
{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
```

Bodigrim commented:
@sjakobi sorry, I lost track of the discussion. Is there anything left uncovered?

sjakobi commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,269 @@
+-- "9.999999999999999e22"
+--
+-- Simplifying, we can build a shorter, lossless representation by just using
+-- @"1.0e23"@ since the floating point values that are 1 ULP away are
```
What is "ULP"?

la-wu replied:
[Unit in Last Place](https://en.wikipedia.org/wiki/Unit_in_the_last_place)

sjakobi replied:
Good to know! Could you link to the article?

Note BTW that haddock supports Markdown-style hyperlinks, e.g. `[ULP](https://en.wikipedia.org/wiki/Unit_in_the_last_place)`.

sjakobi commented on Data/ByteString/Builder/RealFloat.hs:
```diff
@@ -0,0 +1,269 @@
+-- output (specifically with respect to default rounding and length). In
+-- particular, there are boundary cases where the closest and \'shortest\'
+-- string representations are not used.  Mentions of \'shortest\' in the docs
+-- below are with this caveat.
+--
+-- For example, for fidelity, we match `show` on the output below.
+--
+-- >>> show (1.0e23 :: Float)
+-- "1.0e23"
+-- >>> show (1.0e23 :: Double)
+-- "9.999999999999999e22"
+-- >>> floatDec 1.0e23
+-- "1.0e23"
+-- >>> doubleDec 1.0e23
+-- "9.999999999999999e22"
+--
+-- Simplifying, we can build a shorter, lossless representation by just using
+-- @"1.0e23"@ since the floating point values that are 1 ULP away are
+--
+-- >>> showHex (castDoubleToWord64 1.0e23) []
+-- "44b52d02c7e14af6"
+-- >>> castWord64ToDouble 0x44b52d02c7e14af5
+-- 9.999999999999997e22
+-- >>> castWord64ToDouble 0x44b52d02c7e14af6
+-- 9.999999999999999e22
+-- >>> castWord64ToDouble 0x44b52d02c7e14af7
+-- 1.0000000000000001e23
+--
+-- In particular, we could use the exact boundary if it is the shortest
+-- representation and the original floating number is even. To experiment with
+-- the shorter rounding, refer to
+-- `Data.ByteString.Builder.RealFloat.Internal.acceptBounds`. This will give us
+--
+-- >>> floatDec 1.0e23
+-- "1.0e23"
+-- >>> doubleDec 1.0e23
+-- "1.0e23"
+--
+-- For more details, please refer to the
+-- <https://dl.acm.org/doi/10.1145/3192366.3192369 Ryu paper>.
```
I think this section is a bit too technical for most users. If we want to keep it, we should probably move it out of the module header and to the bottom of the page.

la-wu replied:
Is there such a thing as a footer? This was added per one of our discussions above about clarifying 'shortest'.

sjakobi replied:
You can insert chunks of documentation at selected places in the export list. See e.g. [`Data.ByteString.Lazy`](https://hackage.haskell.org/package/bytestring-0.11.1.0/docs/Data-ByteString-Lazy.html#g:25).

sjakobi replied:
Also see the relevant haddock documentation at https://haskell-haddock.readthedocs.io/en/latest/markup.html#controlling-the-documentation-structure.

sjakobi reviewed (commented):
I'm happy with the API at this stage.

The haddocks could use some editing and more examples, but I don't think this needs to happen within this PR.

The GHC warning should be addressed before merging though.

Bodigrim commented:
@sjakobi thanks for extensive reviews! 

@la-wu if you get inspiration to fix documentation suggestions in a separate PR, it would be awesome. Otherwise thank you very much for all your heroic efforts to make this state-of-the-art algorithm available to Haskell users. This is a long-awaited improvement.

sjakobi commented:
@la-wu Thanks for enduring this marathon of a review! :)

You may want to add your implementation to the list at https://github.com/ulfjack/ryu#ryu, BTW!

hsyl20 commented:
@la-wu Out of curiosity, did you make benchmarks against the `double-conversion` package? [We have had troubles using this package with GHCJS recently (it needs to be linked with libstdc++...) so a pure Haskell solution is very appealing if performance is comparable]

la-wu commented:
> @la-wu Out of curiosity, did you make benchmarks against the `double-conversion` package? [We have had troubles using this package with GHCJS recently (it needs to be linked with libstdc++...) so a pure Haskell solution is very appealing if performance is comparable]

Hey @hsyl20, not sure if this is still relevant but I just ran a quick bench for this and I got

```
foldMap floatDec (10000):                               OK (0.15s)
  1.17 ms ± 114 μs
foldMap doubleDec (10000):                              OK (0.37s)
  1.42 ms ±  45 μs
foldMap floatShow (10000):                              OK (0.33s)
  10.5 ms ± 541 μs
foldMap doubleShow (10000):                             OK (0.27s)
  18.0 ms ± 1.0 ms
```

And with double-conversion,

```
foldMap floatDC (10000):    OK (0.16s)
  5.17 ms ± 450 μs
foldMap doubleDC (10000):   OK (0.21s)
  6.55 ms ± 494 μs
```

With a setup of

```
import qualified Data.Double.Conversion.Convertable    as DC

...
, benchB "foldMap floatShow"       floatData          $ foldMap (string7 . show)
, benchB "foldMap doubleShow"      doubleData         $ foldMap (string7 . show)
, benchB "foldMap floatDC"         floatData          $ foldMap DC.toShortest
, benchB "foldMap doubleDC"        doubleData         $ foldMap DC.toShortest
```

Not sure if I was using the library correctly... but directly converting with double-conversion to Bytestring (instead of builder) was much slower fwiw.

hsyl20 commented:
@la-wu Thanks a lot for this report! We should definitely use your code instead of having to deal with `double-conversion`. I've added it to our task list.
